force_bdss/data_sources/data_source_utilities.py

```diff
@@ -0,0 +1,196 @@
+    source, target: objects
+        Object instances to perform attribute checks on
+    attributes: str or list of str
+        An attribute or list of attributes used to check merge
+        eligibility of source and target objects
+    ignore_default: bool, optional, default: False
+        Whether or not to ignore any attributes on source or target
+        that are are equal to their default values
+
+    Returns
+    -------
+    failed_attr: list of str
+        Names of attributes on source and target that fail a
+        merge check
+    """
+
+    # Ensure attributes argument is in an iterable format
+    if attributes is None:
+        attributes = []
+    elif isinstance(attributes, str):
+        attributes = [attributes]
+    else:
+        attributes = list(attributes)
+
+    failed_attr = []
+
+    # Iterate through attributes list
+    for attr_name in attributes:
+        if not trait_similarity_check(
+                source, target,
+                attr_name, ignore_default):
+            failed_attr.append(attr_name)
+
+    return failed_attr
+
+
+def merge_trait_with_check(source, target, name, attributes=None,
+                           ignore_default=False):
+    """Performs the same function as `merge_trait` method, but with
+    some extra logic checks on default values and any selected
+    attributes supplied by `attributes` argument. These checks are
+    designed to determine whether both objects possess a similar
+    enough state to merge the `name` attribute.
+
+    Parameters
+    ----------
+    source, target: HasTraits
+        HasTraits instances to have an attribute synchronised
+    name: str
+        Name of an attribute on both source and target to be synced
+    attributes: str or list of str, optional
+        An attribute or list of attributes to check equivalence on both
+        source and target objects
+    ignore_default: bool, optional, default: False
+        Whether or not to ignore any attributes on source or target
+        that are are equal to their default values
+
+    Raises
+    -----
+    TraitSimilarityError, if any attribute checks fail
+    """
+
+    # Obtain names of any attributes in `attr_check` on both source and
+    # target that fail a `trait_similarity_check`
+    failed_attr = attr_checker(source, target, attributes,
+                               ignore_default=ignore_default)
+
+    if failed_attr:
+        attr_name = failed_attr[0]
+        error_msg = (
+            "Source object has failed a trait "
+            "similarity check with target: "
+            "The {} attribute of source ({}) doesn't match "
+            "target ({}).".format(
+                attr_name,
+                getattr(source, attr_name),
+                getattr(target, attr_name)
+            )
+        )
+        logger.exception(error_msg)
+        raise TraitSimilarityError(error_msg)
+
+    # Merge attribute `name` between source and target
+    merge_trait(source, target, name)
```

Comment:
Would it make sense to refactor this function to separate the `check` part from the `merge_trait` part? Right now it seems that we are doing two things at a time. Maybe it could only perform the `if failed_attr: ...` verification, and return nothing if everything is fine. And then, in the method where `merge_trait_with_check` is called, we have two lines: one for the check, and the second for `merge_trait`.